api/v3/Case.php

```diff
@@ -60,58 +62,117 @@
 
-  // Fixme: can we safely pass raw params to the BAO?
-  $newParams = array(
-    'case_type_id' => $params['case_type_id'],
-    'creator_id' => $params['creator_id'],
-    'status_id' => $params['status_id'],
-    'start_date' => $params['start_date'],
-    'end_date' => CRM_Utils_Array::value('end_date', $params),
-    'subject' => $params['subject'],
-  );
+    // get original contact id and creator id of case
+    if (!empty($params['contact_id'])) {
+      // FIXME: CRM_Case_BAO_Case::retrieveContactIdsByCaseId returns a 1-based array (1 is the first element)
+      // It should really return a 0-based array for consistency.
+      $origContactIds = CRM_Case_BAO_Case::retrieveContactIdsByCaseId($params['id']);
+      $origContactId = $origContactIds[1];
+    }
 
-  $caseBAO = CRM_Case_BAO_Case::create($newParams);
+    // FIXME: Refactor as separate method to get contactId
+    if (count($origContactIds) > 1) {
+      // check valid orig contact id
+      if (empty($params['orig_contact_id'])) {
+        throw new API_Exception('Case is linked with more than one contact id. Provide the required params orig_contact_id to be replaced');
+      }
+      if (!empty($params['orig_contact_id']) && !in_array($params['orig_contact_id'], $origContactIds)) {
+        throw new API_Exception('Invalid case contact id (orig_contact_id)');
+      }
+      $origContactId = $params['orig_contact_id'];
+    }
+
+    // check for same contact id for edit Client
+    if (!empty($params['contact_id']) && !in_array($params['contact_id'], $origContactIds)) {
+      $mergedCaseId = CRM_Case_BAO_Case::mergeCases($params['contact_id'], $params['case_id'], $origContactId, NULL, TRUE);
+    }
+
+    // If we merged cases then update the merged case
+    if (!empty($mergedCaseId[0])) {
+      $params['id'] = $mergedCaseId[0];
+    }
+  }
+
+  // Create/update the case
+  $caseBAO = CRM_Case_BAO_Case::create($params);
 
   if (!$caseBAO) {
     throw new API_Exception('Case not created. Please check input params.');
   }
 
-  foreach ((array) $params['contact_id'] as $cid) {
-    $contactParams = array('case_id' => $caseBAO->id, 'contact_id' => $cid);
-    CRM_Case_BAO_CaseContact::create($contactParams);
+  if (isset($params['contact_id'])) {
+    foreach ((array) $params['contact_id'] as $cid) {
+      $contactParams = array('case_id' => $caseBAO->id, 'contact_id' => $cid);
+      CRM_Case_BAO_CaseContact::create($contactParams);
+    }
+  }
+
+  if (!isset($params['id'])) {
+    // As the API was not passed an id we have created a new case.
+    // Only run the xmlProcessor for new cases to get all configuration for the new case.
+    _civicrm_api3_case_create_xmlProcessor($params, $caseBAO);
+  }
+
+  // return case
+  $values = array();
+  _civicrm_api3_object_to_array($caseBAO, $values[$caseBAO->id]);
+
+  // Add custom data
+  if (isset($params['custom'])) {
+    _civicrm_api3_custom_data_get($values[$caseBAO->id], TRUE, 'case', $caseBAO->id);
+  }
```

Comment:
This is an extra database hit, which we normally don't do for create operations. There is an api option called something like "force_reload" which takes care of this type of thing.